client.go

```diff
@@ -545,6 +553,11 @@ func PassthroughErrorHandler(resp *http.Response, err error, _ int) (*http.Respo
 
 // Do wraps calling an HTTP method with retries.
 func (c *Client) Do(req *Request) (*http.Response, error) {
+	return c.DoWithResponseHandler(req, nil)
+}
+
+// DoWithResponseHandler wraps calling an HTTP method plus a response handler with retries.
+func (c *Client) DoWithResponseHandler(req *Request, handler func(*http.Response) (shouldRetry bool)) (*http.Response, error) {
```

Comment:
Good call, totally agreed on setting it on the Request type.

Re: bool vs error I do think it's important to not _just_ return an error since some errors are retryable and others are not. But I think what I'll do is require it returns both an error and a bool

Reply:
Oh the other difficulty is "handling it all together" is a breaking change since `CheckRetry` is exported. To keep things simple for now I'll avoid modifying its signature.

Reply:
Actually the more I think about this I'm getting my head in circles. Which is fine except if the interface for consumers of the library makes _their_ heads also spin. Let's chat briefly about this over zoom (we can do so at our 1:1 today)